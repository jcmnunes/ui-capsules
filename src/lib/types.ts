import { icons } from './Icon/Icon.constants';
import { illustratedIcons } from './IllustratedIcon/IllustratedIcon.constants';
import { theme } from './theme';
import { solidIcons } from './IconSolid/IconSolid.constants';
import { MouseEvent } from 'react';

export type VariantColor =
  | 'neutral'
  | 'primary'
  | 'info'
  | 'success'
  | 'warning'
  | 'error'
  | 'blue'
  | 'indigo'
  | 'purple'
  | 'pink'
  | 'red'
  | 'orange'
  | 'yellow'
  | 'green'
  | 'teal';

type ActionVariant = 'primary' | 'neutral' | 'error';

export type Action = {
  text: string;
  variant?: ActionVariant;
  onClick(e: MouseEvent<HTMLButtonElement>): void;
};

export type ElementSize = 'small' | 'medium' | 'large';

export type DualSize = 'medium' | 'large';

export type AlignmentOptions = 'left' | 'center' | 'right';

export type IconType = keyof typeof icons;

export type IconSolidType = keyof typeof solidIcons;

export type IconVariant = 'solid' | 'outline';

export type IllustratedIconType = keyof typeof illustratedIcons;

export type Theme = typeof theme;

export type SelectOption<T> = { label: 'string'; value: T };
