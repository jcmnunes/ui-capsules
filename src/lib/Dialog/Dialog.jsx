import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Button from '../Button/Button';
import Modal, { ModalBody, ModalFooter } from '../Modal/Modal';
import theme from '../theme';

export const StyledModal = styled(Modal)`
  width: auto;

  @media (min-width: ${theme.bp_mobile}) {
    left: 60px;
    right: 60px;
  }

  @media (min-width: ${theme.bp_small}) {
    width: 400px;
    left: 50%;
    right: auto;
  }
`;

const Message = styled.h1`
  font-size: 18px;
  color: ${theme.neutral600};
  text-align: left;
`;

const StyledModalFooter = styled(ModalFooter)`
  justify-content: flex-end;
  background: ${theme.neutral050};
`;

const Dialog = ({ isOpen, actions, contentLabel, onRequestClose, children }) => {
  return (
    <StyledModal
      isOpen={isOpen}
      contentLabel={contentLabel}
      onRequestClose={onRequestClose}
      showCloseButton={false}
    >
      <ModalBody>
        <Message>{children}</Message>
      </ModalBody>
      <StyledModalFooter data-testid="dialog-footer">
        {actions.map(({ name, appearance, action }) => (
          <Button key={name} type="button" appearance={appearance} size="large" onClick={action}>
            {name}
          </Button>
        ))}
      </StyledModalFooter>
    </StyledModal>
  );
};
Dialog.displayName = 'Dialog';

Dialog.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  contentLabel: PropTypes.string.isRequired,
  actions: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      intent: PropTypes.string,
      action: PropTypes.func.isRequired,
    }),
  ).isRequired,
  onRequestClose: PropTypes.func.isRequired,
  children: PropTypes.string.isRequired,
};

export default Dialog;
