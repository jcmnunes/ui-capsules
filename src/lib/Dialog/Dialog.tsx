import React, { FC } from 'react';
import styled from 'styled-components';
import ReactModal from 'react-modal';
import { Modal, ModalFooter, ModalBody, ModalTitle } from '../Modal/Modal';
import { Button } from '../Button/Button';
import { theme } from '../theme';
import { DialogAction } from '../types';

const StyledModal = styled(Modal)`
  width: auto;

  @media (min-width: ${theme.breakpointsLegacy.mobile}) {
    left: 60px;
    right: 60px;
  }

  @media (min-width: ${theme.breakpointsLegacy.sm}) {
    width: 450px;
    left: 50%;
    right: auto;
  }
`;

const Message = styled.p`
  font-size: 16px;
  color: ${theme.neutral600};
  text-align: left;
`;

const StyledModalFooter = styled(ModalFooter)`
  justify-content: flex-end;
  background: ${theme.neutral050};
`;

const StyledModalBody = styled(ModalBody)<{ hasTitle: boolean }>`
  padding-top: ${({ hasTitle }) => hasTitle && '12px'};
`;

interface Props extends ReactModal.Props {
  isOpen: boolean;
  contentLabel: string;
  actions: DialogAction[];
  onRequestClose(): void;
  title?: string;
  message: string;
}

export const Dialog: FC<Props> = ({
  isOpen,
  actions,
  contentLabel,
  onRequestClose,
  title,
  message,
  ...other
}) => {
  return (
    <StyledModal
      title="Example modal"
      isOpen={isOpen}
      contentLabel={contentLabel}
      onRequestClose={onRequestClose}
      showCloseButton={false}
      {...other}
    >
      <StyledModalBody hasTitle={!!title}>
        {!!title && <ModalTitle>{title}</ModalTitle>}
        <Message>{message}</Message>
      </StyledModalBody>
      <StyledModalFooter data-testid="dialog-footer">
        {actions.map(({ name, action }) => (
          <Button key={name} type="button" size="large" onClick={action}>
            {name}
          </Button>
        ))}
      </StyledModalFooter>
    </StyledModal>
  );
};
Dialog.displayName = 'Dialog';
