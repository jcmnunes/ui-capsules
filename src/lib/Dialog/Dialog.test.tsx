import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import RModal from 'react-modal';
import { Dialog } from './Dialog';
import { DialogAction } from '../types';

const bindApp = () => {
  const app = document.createElement('div');
  app.setAttribute('id', 'root');
  document.body.append(app);
  RModal.setAppElement('#root');
};

const CONTENT_LABEL = 'Example Dialog';
const DIALOG_TEXT = 'Are you sure you want to delete?';

const getDialogElement = (baseElement: HTMLElement) =>
  baseElement.querySelector(`[aria-label="${CONTENT_LABEL}"]`);

beforeEach(() => {
  bindApp();
});

afterEach(() => {
  document.body.removeChild(document.querySelector('#root') as Node);
});

it('should have a displayName', () => {
  expect(Dialog.displayName).toBe('Dialog');
});

describe('Dialog with two actions', () => {
  const cancelAction = jest.fn();
  const deleteAction = jest.fn();
  const actions: DialogAction[] = [
    {
      name: 'Cancel',
      appearance: 'secondary',
      action: cancelAction,
    },
    {
      name: 'Delete',
      appearance: 'error',
      action: deleteAction,
    },
  ];

  // eslint-disable-next-line react/prop-types
  const DialogTwoActions = ({ isOpen }: { isOpen: boolean }) => (
    <Dialog
      isOpen={isOpen}
      onRequestClose={cancelAction}
      contentLabel="Example Dialog"
      actions={actions}
      message={DIALOG_TEXT}
    />
  );

  beforeEach(() => {
    cancelAction.mockReset();
    deleteAction.mockReset();
  });

  it('should render with no errors', () => {
    const { baseElement, getByText } = render(<DialogTwoActions isOpen />);
    expect(getByText(DIALOG_TEXT)).toBeInTheDocument();
    expect(getDialogElement(baseElement)).toMatchSnapshot();
  });

  it('should not render the dialog when isOpen is set to false', () => {
    const { baseElement } = render(<DialogTwoActions isOpen={false} />);
    expect(getDialogElement(baseElement)).toBeNull();
  });

  it('should set the aria-label attribute', () => {
    const { baseElement } = render(<DialogTwoActions isOpen />);
    expect(getDialogElement(baseElement)).toHaveAttribute('aria-label', CONTENT_LABEL);
  });

  it('should call actions when clicking the footer buttons (1)', () => {
    const { getByText } = render(<DialogTwoActions isOpen />);
    fireEvent.click(getByText('Cancel'));
    expect(cancelAction).toHaveBeenCalledTimes(1);
    expect(deleteAction).toHaveBeenCalledTimes(0);
  });

  it('should call actions when clicking the footer buttons (2)', () => {
    const { getByText } = render(<DialogTwoActions isOpen />);
    fireEvent.click(getByText('Delete'));
    expect(cancelAction).toHaveBeenCalledTimes(0);
    expect(deleteAction).toHaveBeenCalledTimes(1);
  });

  it('should render one button per action', () => {
    const { getByTestId } = render(<DialogTwoActions isOpen />);
    const footer = getByTestId('dialog-footer');
    const buttons = footer.querySelectorAll('button');
    expect(buttons.length).toBe(actions.length);
  });
});
