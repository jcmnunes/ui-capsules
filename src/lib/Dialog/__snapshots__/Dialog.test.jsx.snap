// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog with two actions should render with no errors 1`] = `
.c4 {
  visibility: visible;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c9 {
  white-space: nowrap;
  margin-left: auto;
  margin-right: auto;
}

.c3 {
  position: relative;
  height: 32px;
  color: inherit;
  border-radius: 4px;
  border: none;
  border: none;
  font-size: 16px;
  font-family: 'Source Sans Pro',Helvetica,Arial,sans-serif;
  padding: 4px 6px;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: auto;
  background: linear-gradient(to bottom, none, none );
}

.c3::-webkit-input-placeholder {
  color: #9AA5B1;
}

.c3::-moz-placeholder {
  color: #9AA5B1;
}

.c3:-ms-input-placeholder {
  color: #9AA5B1;
}

.c3::placeholder {
  color: #9AA5B1;
}

.c3:hover {
  background: linear-gradient(to bottom, none, none );
}

.c3:active {
  background: none;
}

.c3:disabled {
  background: none;
  cursor: not-allowed;
}

.c8 {
  position: relative;
  height: 48px;
  color: #52616D;
  border-radius: 4px;
  border: none;
  border: 1px solid #C0C2C6;
  font-size: 18px;
  font-family: 'Source Sans Pro',Helvetica,Arial,sans-serif;
  padding: 12px;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: auto;
  background: linear-gradient(to bottom, #F5F7FA, #E4E7EB );
}

.c8::-webkit-input-placeholder {
  color: #9AA5B1;
}

.c8::-moz-placeholder {
  color: #9AA5B1;
}

.c8:-ms-input-placeholder {
  color: #9AA5B1;
}

.c8::placeholder {
  color: #9AA5B1;
}

.c8:hover {
  background: linear-gradient(to bottom, #E2E7EC, #D7DBE0 );
}

.c8:active {
  background: #CBD2D8;
}

.c8:disabled {
  background: #F5F7FA;
  cursor: not-allowed;
}

.c10 {
  position: relative;
  height: 48px;
  color: #FFF;
  border-radius: 4px;
  border: none;
  border: none;
  font-size: 18px;
  font-family: 'Source Sans Pro',Helvetica,Arial,sans-serif;
  padding: 12px;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: auto;
  background: linear-gradient(to bottom, #E12D39, #CF1124 );
}

.c10::-webkit-input-placeholder {
  color: #9AA5B1;
}

.c10::-moz-placeholder {
  color: #9AA5B1;
}

.c10:-ms-input-placeholder {
  color: #9AA5B1;
}

.c10::placeholder {
  color: #9AA5B1;
}

.c10:hover {
  background: linear-gradient(to bottom, #CF1124, #AB091E );
}

.c10:active {
  background: #AB091E;
}

.c10:disabled {
  background: #FFBCC2;
  cursor: not-allowed;
}

.c1 {
  position: absolute;
  top: 50%;
  left: 16px;
  right: 16px;
  bottom: auto;
  -webkit-transform: translate(0,-50%);
  -ms-transform: translate(0,-50%);
  transform: translate(0,-50%);
  background: #fff;
  overflow: auto;
  -webkit-overflow-scrolling: touch;
  outline: none;
  width: auto;
  min-width: 300px;
  box-shadow: rgba(0,0,0,0.15) 0 10px 20px,rgba(0,0,0,0.1) 0 3px 6px;
  border-width: 6px;
  border-style: solid;
  border-color: #CBD2D8;
  border-image: initial;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  height: 40px;
  padding: 16px;
  color: #CBD2D8;
}

.c5 {
  padding: 12px 32px 32px;
}

.c0 {
  width: auto;
}

.c6 {
  font-size: 18px;
  color: #52616D;
  text-align: center;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  background: #F5F7FA;
  padding: 18px 24px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background: #fff;
}

.c7 > * {
  margin-left: 16px;
}

@media (min-width:450px) {
  .c1 {
    left: 40px;
    right: 40px;
  }
}

@media (min-width:720px) {
  .c1 {
    top: 50%;
    left: 50%;
    right: auto;
    bottom: auto;
    margin-right: -50%;
    width: 640px;
    -webkit-transform: translate(-50%,-50%);
    -ms-transform: translate(-50%,-50%);
    transform: translate(-50%,-50%);
  }
}

@media (min-width:450px) {
  .c0 {
    left: 60px;
    right: 60px;
  }
}

@media (min-width:720px) {
  .c0 {
    width: 400px;
    left: 50%;
    right: auto;
  }
}

<div
  aria-label="Example Dialog"
  class="ReactModal__Content ReactModal__Content--after-open c0 c1"
  role="dialog"
  tabindex="-1"
>
  <div
    class="c2"
  >
    <button
      class="c3"
    >
      <div
        class="c4"
      >
        <svg
          viewBox="0 0 20 20"
          width="20px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11.88 10l5.73-5.71a1.34 1.34 0 1 0-1.9-1.9L10 8.12 4.29 2.39a1.34 1.34 0 0 0-1.9 1.9L8.12 10l-5.73 5.71a1.34 1.34 0 0 0 0 1.89 1.34 1.34 0 0 0 1.89 0L10 11.88l5.71 5.73a1.34 1.34 0 0 0 1.89 0 1.34 1.34 0 0 0 0-1.89z"
            fill="currentColor"
          />
        </svg>
      </div>
    </button>
  </div>
  <div
    class="c5"
  >
    <h1
      class="c6"
    >
      Are you sure you want to delete?
    </h1>
  </div>
  <div
    class="c7"
    data-testid="dialog-footer"
  >
    <button
      class="c8"
    >
      <div
        class="c4"
      >
        <span
          class="c9"
        >
          Cancel
        </span>
      </div>
    </button>
    <button
      class="c10"
    >
      <div
        class="c4"
      >
        <span
          class="c9"
        >
          Delete
        </span>
      </div>
    </button>
  </div>
</div>
`;
